### ISSUE TITLE
lint_serial_number_not_positive does not flag serial numbers == 0

### ISSUE THREAD

**Issue by CBonnell** [zmap/zlint#582]
https://github.com/zmap/zlint/blob/master/v3/lints/rfc/lint_serial_number_not_positive.go#L62 only flags an error if the certificate serial number is negative. RFC 5280 requires that the serial number be positive, which excludes "0" as a valid value.

See https://crt.sh/?id=12724095&opt=cablint,zlint as an example where cablint correctly flags this error condition but zlint does not.

**Comment by christopher-henderson** [comment]
Hmmm we might actually be the ones who are correct here.

> Appendix B.  ASN.1 Notes
>
>   CAs MUST force the serialNumber to be a non-negative integer, that
>   is, the sign bit in the DER encoding of the INTEGER value MUST be
>   zero.  This can be done by adding a leading (leftmost) `00'H octet if
>  necessary.  This removes a potential ambiguity in mapping between a
>   string of octets and an integer value.

And checking the sign in this fashion is exactly what we do...

```go
if cert.SerialNumber.Sign() == -1 { // -1 Means negative when using big.Sign()
	return &lint.LintResult{Status: lint.Error}
} else {
	return &lint.LintResult{Status: lint.Pass}
}
```

However, what is vexing is the following clause from 4.1.2.2

>    Note: Non-conforming CAs may issue certificates with serial numbers
>   that are negative or zero.  Certificate users SHOULD be prepared to
>  gracefully handle such certificates.

...which seems to imply that the intent of the author of this paragraph is that zeroes are _not_ legal.

Could I ask for a favor from @sleevi  and weigh in on this? You tend to have a much better pulse on the original intent of these documents than I do, and there appears to be some cognitive dissonance in 5280 with regard to this.

**Comment by christopher-henderson** [comment]
Hmmm lets take a look at this from a number line perspective

![image](https://user-images.githubusercontent.com/8518514/113201165-11c92080-921e-11eb-84f6-4bd9136161fe.png)

Zero is commonly considered neither positive nor negative. So you can think of the `positive` set, the `negative` set, and the `zero` singleton set. In this take

`positive` --> `positive`
`negative` --> `negative`
`non-negative` -> `zero` U `positive`
`non-positive` - > `zero` U `negative`

Unfortunately 4.1.2.2 states that...

> The serial number MUST be a positive integer

...and the appendix is of the opinion that...

> CAs MUST force the serialNumber to be a non-negative integer

:man_shrugging:

**Comment by CBonnell** [comment]
Obviously I am not @sleevi, but my understanding about 0 being prohibited is based on the specification in https://tools.ietf.org/html/rfc5280#section-4.1.2.2, specifically:

> The serial number MUST be a positive integer assigned by the CA to
> each certificate...
> Note: Non-conforming CAs may issue certificates with serial numbers
> that are negative or zero.  Certificate users SHOULD be prepared to
> gracefully handle such certificates.

This exact issue was discussed several years ago on m.d.s.p. and the conclusion is that 0 is prohibited: https://groups.google.com/g/mozilla.dev.security.policy/c/3avqmSF4MVU/m/ZPFE0rIuAQAJ

**Comment by sleevi** [comment]
@christopher-henderson I'm not seeing the dissonance. Could you spell it out a bit more?

Appendix B is talking about the **sign bit**, not the entire serial. Prefixing a serial with `00` resolves the sign-bit, but it's not the same as encouraging an all-zero serial.

**Comment by christopher-henderson** [comment]
@sleevi Specifically because one uses `positive` and the other uses `non-negative`, for which the latter does include zero.

> The serial number MUST be a positive integer

> CAs MUST force the serialNumber to be a non-negative integer

Although I would say that my take would be more convincing if instead the Appendix said

> CAs MUST issue a serialNumber that is a non-negative integer

That is, I would accept that the Appendix is just giving a hot tip that the leading ASN,1 encoding has to be zero and to be careful that your number (say, `0011`) does not get encoded with the leading zeroes stripped away (`11`).

**Comment by sleevi** [comment]
Ah, yeah, gotcha. As the thread @CBonnell linked to covered, the reading of Appendix B is "negative or not-negative", since it's talking about sign-bit, while the reading of 4.1.2.2 is "positive or not-positive. But also, not zero".

But yeah, a sole `00` is nonsense.

**Comment by christopher-henderson** [comment]
:man_shrugging: I scanned the conversation, but apparently did so too fast and missed that resolution. Thank you!
